File: packages/flutter_tools/gradle/src/test/kotlin/FlutterExtensionTest.kt
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.flutter.gradle

import org.gradle.api.GradleException
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class FlutterExtensionTest {
    @Test
    fun `getVersionCode() throws GradleException when flutterVersion is not set`() {
        val flutterExtension: FlutterExtension = FlutterExtension()
        assertFailsWith<GradleException> { flutterExtension.getVersionCode() }
    }

    @Test
    fun `getVersionCode() throws GradleException when flutterVersion is not an integer`() {
        val flutterExtension: FlutterExtension = FlutterExtension()
        flutterExtension.flutterVersionCode = "not an integer"
        assertFailsWith<GradleException> { flutterExtension.getVersionCode() }
    }

    @Test
    fun `getVersionCode() returns flutterVersion without error when set and is a number`() {
        val flutterExtension: FlutterExtension = FlutterExtension()
        flutterExtension.flutterVersionCode = "123"
        assertEquals(123, flutterExtension.getVersionCode())
    }

    @Test
    fun `getVersionName() throws GradleException when flutterVersionName is not set`() {
        val flutterExtension: FlutterExtension = FlutterExtension()
        assertFailsWith<GradleException> { flutterExtension.getVersionName() }
    }

    @Test
    fun `getVersionName() returns flutterVersionName without error when set`() {
        val flutterExtension: FlutterExtension = FlutterExtension()
        flutterExtension.flutterVersionName = "1.2.3"
        assertEquals("1.2.3", flutterExtension.getVersionName())
    }
}

File: packages/flutter_tools/test/general.shard/isolated/flutter_extension_test.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:file/memory.dart';
import 'package:flutter_hook_config/flutter_hook_config.dart';
import 'package:flutter_tools/src/artifacts.dart';
import 'package:flutter_tools/src/base/file_system.dart';
import 'package:flutter_tools/src/isolated/native_assets/native_assets.dart';

import '../../src/common.dart';

void main() {
  late FileSystem fileSystem;
  late Artifacts artifacts;

  setUp(() {
    fileSystem = MemoryFileSystem.test();
    artifacts = Artifacts.test(fileSystem: fileSystem);
  });

  testWithoutContext('createFlutterExtension resolves host tools to absolute file URIs', () {
    final FlutterExtension extension = createFlutterExtension(
      artifacts: artifacts,
      engineVersion: 'abc123',
    );

    expect(extension.engineVersion, 'abc123');
    expect(extension.impellerc.isScheme('file'), true);
    expect(extension.impellerc.isAbsolute, true);
    expect(
      extension.impellerc,
      Uri.file(artifacts.getHostArtifact(HostArtifact.impellerc).absolute.path),
    );
    expect(extension.libtessellator.isScheme('file'), true);
    expect(extension.libtessellator.isAbsolute, true);
    expect(
      extension.libtessellator,
      Uri.file(artifacts.getHostArtifact(HostArtifact.libtessellator).absolute.path),
    );
  });

  testWithoutContext('local engine version is stable until a host tool changes', () {
    final File impellerc = fileSystem.file(artifacts.getHostArtifact(HostArtifact.impellerc).path)
      ..writeAsStringSync('impellerc binary');
    fileSystem
        .file(artifacts.getHostArtifact(HostArtifact.libtessellator).path)
        .writeAsStringSync('libtessellator binary');

    final String before = createFlutterExtension(
      artifacts: artifacts,
      engineVersion: null,
    ).engineVersion;
    expect(before, isNotEmpty);
    expect(createFlutterExtension(artifacts: artifacts, engineVersion: null).engineVersion, before);

    impellerc.writeAsStringSync('rebuilt impellerc binary');
    final String after = createFlutterExtension(
      artifacts: artifacts,
      engineVersion: null,
    ).engineVersion;
    expect(after, isNot(before));
  });
}
